telemetry: report segment events asynchronously

gcurtis opened the pull request:
Instead of blocking the process from exiting, buffer Segment events to disk and start another process to upload them. With this change all telemetry should be non-blocking.

Most of the Segment event logic now lives in the telemetry package instead of the cobra middleware. A bunch of the unused logic around tracking CLI SSH sessions has also been removed. The telemetry package API now looks something like:

```go
type EventName int

const (
	EventCommandSuccess EventName = iota
	EventShellInteractive
	EventShellReady
)

func Error(err error, meta Metadata)
func Event(e EventName, meta Metadata)
func Start()
func Stop()
func Upload()
```

Fixes #1003.

savil commented on internal/envir/util.go:
```diff
@@ -43,6 +39,28 @@ func IsCI() bool {
 	return ci && err == nil
 }
 
+func ShellStart() time.Time {
+	return ParseShellStart(os.Getenv(DevboxShellStartTime))
+}
+
+func FormatShellStart(t time.Time) string {
+	if t.IsZero() {
+		return ""
+	}
+	return strconv.FormatInt(t.Unix(), 10)
+}
+
+func ParseShellStart(s string) time.Time {
+	if s == "" {
+		return time.Time{}
+	}
+	unix, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return time.Time{}
+	}
+	return time.Unix(unix, 0)
+}
```
I'd prefer keeping this logic out of the `envir` package to keep it streamlined. 

I'd suggest it is better encapsulated in the telemetry package, or even a more focussed `shtelemetry` package

savil commented on internal/telemetry/telemetry.go:
```diff
@@ -4,33 +4,326 @@
+	// Report errors in a separate process so we don't block exiting.
+	exe, err := os.Executable()
+	if err == nil {
+		_ = exec.Command(exe, "bug").Start()
```
could we rename `bug` to be more self-explanatory like `upload-telemetry` ?

savil reviewed (approved):
awesome!

ibuildthecloud commented:
FWIW, this completely breaks in my env :)

I wanted to test this because the telemetry is completely killing me.  You never knew how much a couple 100ms can drive you nuts.

Anyhow.... This hangs forever

```
 cd -
/home/darren/src/oss/devbox
direnv: loading ~/src/oss/devbox/.envrc
direnv: using devbox
direnv: ([/home/linuxbrew/.linuxbrew/Cellar/direnv/2.32.3/bin/direnv export bash]) is taking a while to execute. Use CTRL-C to give up.
```
I'm debugging why.  It's odd because if I run things manually it seems all happy.  To be continued....

ibuildthecloud commented:
This seems suspicious... I have a bunch of `/home/darren/src/oss/devbox/dist/devbox bug` processes laying around.  :thinking:

ibuildthecloud commented:
And this commit definitely breaks it.  I just tested if this is a just a regression else  so I rebased on main.

ibuildthecloud commented:
Ok, it's definitely that.  The `devbox bug` command that is getting spawned is blocking.  I'll let you debug why, I think just doing any `direnv` will reproduce this.  If I have to guess why this is a happening I would say that the `devbox bug` process is stuck on reading stdin or writing stdout/stderr since it's being ran as a subshell in direnv and direnv probably blocks all I/O interaction from the shell.

gcurtis commented:
@ibuildthecloud thanks for catching that. It turns out this was happening because a goroutine was returning without calling `WaitGroup.Done` when the telemetry key wasn't set (which usually only happens in dev builds). I fixed that bug and added a fallback that nukes the process after 5 seconds to make sure we don't leave processes laying around.

ibuildthecloud commented:
Cool this works now (sorta/maybe).  I just now realized that dev build probably don't have telemetry enabled so I don't really know if this is now faster.
